apps/emqx/src/emqx_logger_jsonfmt.erl

```diff
@@ -232,6 +234,15 @@ best_effort_json_obj(Map, Config) ->
             do_format_msg("~p", [Map], Config)
     end.
 
+check_no_dup_tuple_list(List) ->
+    %% Crash if this is not a tuple list
+    lists:foreach(fun({_, _}) -> ok end, List),
+    Items = [K || {K, _} <- List],
+    NumberOfItems = length(Items),
+    %% Crash if there are duplicates
+    NumberOfItems = maps:size(maps:from_keys(Items, true)),
+    ok.
```

Comment:
should avoid building new lists and new maps.

```
to_json_map([], Map) -> Map;
to_json_map([{Key, Value} | More], Map) when is_map_key(Key, Map) ->
    throw(key_duplication);
to_json_map([{Key, Value} | More], Map) ->
   Map#{Key => Value}.
```

Reply:
I think I understand how you are thinking but I don't think your suggestion is the best if one want to avoid creating garbage for the garbage collector to take care of (each evaluation of Map#{Key => Value} will result in a new map on the heap). `maps:from_list/1` is implemented in C and is "quite optimal" from both a performance and garbage generation perspective. I will make a new commit with a more efficient solution.

Reply:
I think this is better:

https://github.com/emqx/emqx/pull/13053/commits/b7c2f4a6d704c35ad9ca8d7b37c8228d622f066f

Thanks for the suggestion!